### ISSUE TITLE
Bought contract assets, immediately sold

### ISSUE THREAD

**Issue by WinPoss** [freqtrade/freqtrade#9460]
## Describe your environment

* Operating system: ubuntu 20.4
* Python Version: 3.9
* CCXT version: 4.0.105
* Freqtrade Version: 2023.10
  
## Your question

*Ask the question you have not been able to find an answer in the [Documentation](https://www.freqtrade.io/en/latest/)*

When my bot bought WLD/USDT:USDT and ARK/USDT:USDT, it immediately closed the position.

May I know the reason for this?

The following is the log content, indicating that the accuracy is not enough:
![log](https://github.com/freqtrade/freqtrade/assets/1667248/68b087aa-bab1-4561-9f2f-2bfd705fcda9)

**Comment by WinPoss** [comment]
```
freqtrade  | 2023-11-22 09:05:11,888 - freqtrade.freqtradebot - INFO - Order 1956668815 was created for WLD/USDT:USDT and status is closed.
freqtrade  | 2023-11-22 09:05:13,750 - freqtrade.wallets - INFO - Wallets synced.
freqtrade  | 2023-11-22 09:05:13,756 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'trade_id': 18, 'type': entry, 'buy_tag': None, 'enter_tag': None, 'exchange': 'Binance', 'pair': 'WLD/USDT:USDT', 'leverage': 1.0, 'direction': 'Long', 'limit': 2.3966, 'open_rate': 2.3966, 'order_type': 'market', 'stake_amount': 86.2776, 'stake_currency': 'USDT', 'base_currency': 'WLD', 'fiat_currency': 'USD', 'amount': 36.0, 'open_date': datetime.datetime(2023, 11, 22, 9, 5, 11, 889060, tzinfo=datetime.timezone.utc), 'current_rate': 2.3966, 'sub_trade': False}
freqtrade  | 2023-11-22 09:05:13,989 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=18, pair=WLD/USDT:USDT, amount=36.00000000, is_short=False, leverage=1.0, open_rate=2.39660000, open_since=2023-11-22 09:05:11)
freqtrade  | 2023-11-22 09:05:14,374 - freqtrade.persistence.trade_model - INFO - Updating trade (id=18) ...
freqtrade  | 2023-11-22 09:05:14,374 - freqtrade.persistence.trade_model - INFO - MARKET_BUY has been fulfilled for Trade(id=18, pair=WLD/USDT:USDT, amount=36.00000000, is_short=False, leverage=1.0, open_rate=2.39660000, open_since=2023-11-22 09:05:11).
freqtrade  | 2023-11-22 09:05:15,086 - freqtrade.wallets - INFO - Wallets synced.
freqtrade  | 2023-11-22 09:05:15,090 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'trade_id': 18, 'type': entry_fill, 'buy_tag': None, 'enter_tag': None, 'exchange': 'Binance', 'pair': 'WLD/USDT:USDT', 'leverage': 1.0, 'direction': 'Long', 'limit': 2.3966, 'open_rate': 2.3966, 'order_type': 'market', 'stake_amount': 86.2776, 'stake_currency': 'USDT', 'base_currency': 'WLD', 'fiat_currency': 'USD', 'amount': 36.0, 'open_date': datetime.datetime(2023, 11, 22, 9, 5, 11, 889060, tzinfo=datetime.timezone.utc), 'current_rate': 2.3966, 'sub_trade': False}
freqtrade  | 2023-11-22 09:05:15,090 - freqtrade.rpc.telegram - INFO - Notification 'entry_fill' not sent.
freqtrade  | 2023-11-22 09:05:15,832 - freqtrade.freqtradebot - ERROR - Unable to place a stoploss order on exchange. Could not create stop_market sell order on market WLD/USDT:USDT. Tried to sell amount 36.0 at rate None. Message: binance price of WLD/USDT:USDT must be greater than minimum price precision of 4
freqtrade  | 2023-11-22 09:05:15,832 - freqtrade.freqtradebot - WARNING - Exiting the trade forcefully
freqtrade  | 2023-11-22 09:05:17,700 - freqtrade.wallets - INFO - Wallets synced.
freqtrade  | 2023-11-22 09:05:17,889 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': exit, 'trade_id': 18, 'exchange': 'Binance', 'pair': 'WLD/USDT:USDT', 'leverage': 1.0, 'direction': 'Long', 'gain': 'loss', 'limit': 0.0, 'order_rate': 0.0, 'order_type': 'market', 'amount': 36.0, 'open_rate': 2.3966, 'close_rate': 0.0, 'current_rate': 2.3966, 'profit_amount': -86.31643182, 'profit_ratio': -1.0, 'buy_tag': None, 'enter_tag': None, 'sell_reason': 'emergency_exit', 'exit_reason': 'emergency_exit', 'open_date': datetime.datetime(2023, 11, 22, 9, 5, 11, 889060, tzinfo=datetime.timezone.utc), 'close_date': datetime.datetime(2023, 11, 22, 9, 5, 17, 889204, tzinfo=datetime.timezone.utc), 'stake_amount': 86.2776, 'stake_currency': 'USDT', 'base_currency': 'WLD', 'fiat_currency': 'USD', 'sub_trade': False, 'cumulative_profit': 0.0}
freqtrade  | 2023-11-22 09:05:17,889 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=18, pair=WLD/USDT:USDT, amount=36.00000000, is_short=False, leverage=1.0, open_rate=2.39660000, open_since=2023-11-22 09:05:11)
freqtrade  | 2023-11-22 09:05:18,560 - freqtrade.persistence.trade_model - INFO - Updating trade (id=18) ...
freqtrade  | 2023-11-22 09:05:18,561 - freqtrade.persistence.trade_model - INFO - MARKET_SELL has been fulfilled for Trade(id=18, pair=WLD/USDT:USDT, amount=36.00000000, is_short=False, leverage=1.0, open_rate=2.39660000, open_since=2023-11-22 09:05:11).
freqtrade  | 2023-11-22 09:05:18,561 - freqtrade.persistence.trade_model - INFO - Marking Trade(id=18, pair=WLD/USDT:USDT, amount=36.00000000, is_short=False, leverage=1.0, open_rate=2.39660000, open_since=closed) as closed as the trade is fulfilled and found no open orders for it.
freqtrade  | 2023-11-22 09:05:18,925 - freqtrade.wallets - INFO - Wallets synced.
```

**Comment by stash86** [comment]
Because the stoploss on exchange failed. Paste your stoploss_on_exchange setting here

**Comment by WinPoss** [comment]
```
stoploss = -0.1

order_types = {
        "entry": "market",
        "exit": "market",
        "emergency_exit": "market",
        "stoploss": "market",
        "stoploss_on_exchange": True,
        "stoploss_on_exchange_interval": 60,
        "stoploss_on_exchange_limit_ratio": 0.99
}
```

**Comment by WinPoss** [comment]
I also set up this function, I don't know if it's the problem with this function?

```
    def custom_stoploss(self, pair: str, trade: 'Trade', current_time: datetime,
                        current_rate: float, current_profit: float, after_fill: bool,
                        **kwargs) -> Optional[float]:

        if current_profit < 0.05:
            return -1

        desired_stoploss = current_profit / 2
        return max(min(desired_stoploss, 0.05), 0.025)
```

**Comment by winposs** [comment]
My trading bot is using the Binance.

**Comment by WinPoss** [comment]
When I set `stoploss_on_exchange` to `False`, I can trade normally. Could it be a bug?

**Comment by ssssi** [comment]
> When I set `stoploss_on_exchange` to `False`, I can trade normally. Could it be a bug?

i had also get emergency_exit in binance due to set stoploss_on_exchange to "True".

**Comment by xmatthias** [comment]
@ssssi without logs, there's nothing to look at, which makes your comment useless, unfortunately.

emergency-exit can have many reasons - most often, it's because of too tight stoplosses configurations, which cause the stoploss placement to fail as price moved past the stop price.

**Comment by ssssi** [comment]
> @ssssi without logs, there's nothing to look at, which makes your comment useless, unfortunately.
> 
> emergency-exit can have many reasons - most often, it's because of too tight stoplosses configurations, which cause the stoploss placement to fail as price moved past the stop price.

You're right, this comment is useless without any proof. This is a problem I encountered a long time ago, the log no longer exists.if encountered again, I will raise the issue.

**Comment by xmatthias** [comment]
> CCXT version: 4.0.105

@winposs your ccxt version seems misaligned. for 2023.10, the correct version is 4.1.22. 4.0.105 is kinda 1 month older. (that's however not the cause for problems i think - as the relevant ccxt code for placing orders didn't change in like 3 months or so).

---

That said - i've just tested exactly your scenario (stop_market order on binance futures) with the latest develop (and uptodate dependencies) - and it works just fine - with parameters similar / identical to yours (limit_rate is None - as the resulting order is a market order).
This was my first suspicion, as some exchanges still require a price for market orders - but binance didn't when i last tested this.

![image](https://github.com/freqtrade/freqtrade/assets/5024695/e8544c92-9ed9-4f38-af52-e4ed750f0e85)

i've used different stop settings than you though, for reasons that will become apparent soon:

---

> I also set up this function, I don't know if it's the problem with this function?
> 
> ```
>     def custom_stoploss(self, pair: str, trade: 'Trade', current_time: datetime,
>                         current_rate: float, current_profit: float, after_fill: bool,
>                         **kwargs) -> Optional[float]:
> 
>         if current_profit < 0.05:
>             return -1
> 
>         desired_stoploss = current_profit / 2
>         return max(min(desired_stoploss, 0.05), 0.025)
> ```


the most likely cause for problems here is the `return -1` statement.

from the [docs](https://www.freqtrade.io/en/stable/strategy-callbacks/#custom-stoploss):

> Returning None will be interpreted as "no desire to change", and is the only safe way to return when you'd like to not modify the stoploss.

My assumption is that below 5% profit, you'd want to keep stoplos at it's initial value (whatever the static `stoploss` defines).

In older versions, the `-1` would've resulted in the stop being ignored, as it's below the original `stoploss` (it's implicitly ignored as it's not allowed to be moved lower than the previous stoploss).
That's also not a breaking change - as it requires explicit opt-in by using the new function signature.

You're however opting into the [new "after_fill " functionality](https://www.freqtrade.io/en/stable/strategy-callbacks/#adjust-stoploss-after-position-adjustments) by having `after_fill: bool,` in the function signature  - which does allow the decrease of stoploss right after entry - accepting any value (including -1). 

Your stoploss will therefore be very low (logs in develop have been enhanced to show the actual stop value) - causing the exchange to reject your order.

As the stoploss can't be placed (but you selected "place stop on exchange") - your trade is at 100% risk (no stoploss could be placed at the desired value) - therefore the bot does the only safe thing - which is exiting the trade immediately.

A safe version would use `return None` instead of the `return -1` statement - or do a condition based on `after_fill` - but i'd reccomend the former, given the intend to "leave stop as is".

**Comment by WinPoss** [comment]
As you mentioned, when I use `return None` in the `custom_stoploss` function, the bot successfully creates a stop-loss order on the Binance platform. Now that I have identified the issue, thank you.

However, I have a question regarding the stop-loss order's stop price:

**When I set the `stoploss_on_exchange_limit_ratio`, is it calculated based on my stop price or the current market price of the asset?**

My `order_types` is:

```
order_types = {
        "entry": "market",
        "exit": "market",
        "emergency_exit": "market",
        "stoploss": "market",
        "stoploss_on_exchange": True,
        "stoploss_on_exchange_interval": 60,
        "stoploss_on_exchange_limit_ratio": 0.99
    }
```

According to the explanation on the official website regarding [stoploss_on_exchange_limit_ratio](https://www.freqtrade.io/en/stable/stoploss/#stoploss_on_exchange-and-stoploss_on_exchange_limit_ratio):

```
Calculation example: we bought the asset at $100.
Stop-price is $95, then limit would be $95 * 0.99 = $94.05 - so the limit order fill can happen between $95 and $94.05.
```

When I bought DYDX/USDT:USDT, the order open price was 3.312 with a short position, and I set `stoploss = -0.1`. Based on my understanding, the stop price should be calculated as: 3.312 * 1.1 = 3.6432, and the limit would be 3.6432 * 0.99 = 3.606768. Therefore, the limit order fill can happen between 3.6432 and 3.606768.

However, in reality, the stop loss price created by the bot on the Binance platform is 4.009.

This price seems to be calculated based on the current market price of DYDX/USDT:USDT, rather than my set stop price. 

Even if I comment out the entire `custom_stoploss` function and set `use_custom_stoploss = False`, and, manually cancel the stop-loss order on the platform, when the bot recreates the stop-loss order, it is not based on the -0.1 stop loss price I set.

Furthermore, when I checked the logs, I noticed that the stop-loss order is automatically canceled and recreated, and the latest stop price also changes.

At this point, I have commented out the `custom_stoploss` function, but set `use_custom_stoploss = True`.

```
freqtrade  | 2023-11-24 10:22:42,461 - freqtrade.worker - INFO - Bot heartbeat. PID=1, version='2023.10', state='RUNNING'
freqtrade  | 2023-11-24 10:22:49,079 - freqtrade.freqtradebot - INFO - Cancelling current stoploss on exchange for pair DYDX/USDT:USDT (orderid:11109738947) in order to add another one ...
freqtrade  | 2023-11-24 10:22:49,079 - freqtrade.freqtradebot - INFO - Canceling stoploss on exchange for Trade(id=33, pair=DYDX/USDT:USDT, amount=26.60000000, is_short=True, leverage=1.0, open_rate=3.31200000, open_since=2023-11-23 18:00:17)
freqtrade  | 2023-11-24 10:22:49,793 - freqtrade.exchange.exchange - INFO - stoploss market order added for DYDX/USDT:USDT. stop price: 3.993. limit: None
freqtrade  | 2023-11-24 10:23:42,464 - freqtrade.worker - INFO - Bot heartbeat. PID=1, version='2023.10', state='RUNNING'
```

**Comment by winposs** [comment]
According to the official example of [custom_stoploss](https://www.freqtrade.io/en/stable/strategy-callbacks/#trailing-stoploss-with-positive-offset):

```
# After reaching the desired offset, allow the stoploss to trail by half the profit
        desired_stoploss = current_profit / 2
```

So, in my function, do I still need to set a `desired_offset = 0.05`?

Thank you.

**Comment by winposs** [comment]
According to the official explanation:

`If stoploss_on_exchange uses limit orders, the exchange needs 2 prices, the stoploss_price and the Limit price.`

So, does it mean that if "entry": "limit" is set, the stop-loss price is calculated based on the opening price of the order (rather than the latest market price), using the set stop-loss ratio, and a stop-loss order is placed on the exchange?

**Comment by WinPoss** [comment]
Why is the stop-loss not set based on the asset's opening price, but rather on the market price?
 I can't figure it out. Is there something I haven't set correctly?

**Comment by xmatthias** [comment]
> When I bought DYDX/USDT:USDT, the order open price was 3.312 with a short position, and I set `stoploss = -0.1`. Based on my understanding, the stop price should be calculated as: 3.312 * 1.1 = 3.6432, and the limit would be 3.6432 * 0.99 = 3.606768. Therefore, the limit order fill can happen between 3.6432 and 3.606768.

This is definitely a wrong understanding (at least for the last part). The example assumes a long trade (which is what is usually easier to understand, as it's more natural for the mind).

the "0.99" will apply in the direction of the stop - so it'd be something around `3.6432 * (2 - 0.99)`.

To return to the documentation sample: 
it's 3 levels - you buy, below that you have the stop price - and again below that, the stop limit price.
**for a long**
Calculation example: we bought the asset at $100.
Stop-price is $95, then limit would be $95 * 0.99 = $94.05 - so the limit order fill can happen between $95 and $94.05.



**for a short**
It's again, 3 levels - where you sell - above that, you have the stop price, and above THAT, you have the stop limit price.

Calculation example: we bought the asset at $100.
Stop-price is $105, then limit would be $105 *(2 -  0.99) ~= $105.95 - so the limit buy order fill can happen between $105 and $105.95.


---

a stoploss will also only update in one direction (and won't update at all if both trailing and custom stoploss are disabled) - so changing stoploss settings on open positions may not yield the actually desired results.

**Comment by WinPoss** [comment]
Thank you for your detailed explanation, I understand now. freqtrade is a fantastic software that has brought me a lot of enjoyment and growth. I will definitely give you a tip once my account starts making profits. Haha!
